core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,169 @@
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(context.getSession(), mergeWriter.getTarget().getHandle());
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+
+        Map<ColumnHandle, Constant> assignments = buildAssignments(orderedColumnNames, mergeAssignments, columnHandles, context);
+        if (assignments.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Query Planner builds the merge rowblock, containing:
+        // All data columns in table order
+        // plus 3 additional fields: field present flag, update operation number, merge case number
+        // TODO move this check before buildAssignments as soon support for update all will be added,
+        //  in case of update all - row encoded as base64, which filtered in buildAssignments
+        checkState(orderedColumnNames.size() == mergeAssignments.size() - 3, "Number of data columns should be equal");
+
+        return metadata.applyUpdate(context.getSession(), tableScan.getTable(), assignments)
+                .map(newHandle -> new TableUpdateNode(
+                        context.getIdAllocator().getNextId(),
+                        newHandle,
+                        getOnlyElement(node.getOutputSymbols())))
+                .map(Result::ofPlanNode)
+                .orElseGet(Result::empty);
+    }
+
+    private Map<ColumnHandle, Constant> buildAssignments(
+            List<String> orderedColumnNames,
+            List<? extends Node> mergeAssignments,
+            Map<String, ColumnHandle> columnHandles,
+            Context context)
+    {
+        ImmutableMap.Builder<ColumnHandle, Constant> assignmentsBuilder = ImmutableMap.builder();
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            if (isSymbol(assigmentNode)) {
+                // the column is not updated
+                continue;
+            }
+            if (!isConstant(assigmentNode)) {
```

Comment:
This is unnecessary, since the call to translate the expression will only produce a Constant if the original expression is constant.